### ISSUE TITLE
Specifying a branch with --git

### ISSUE THREAD

**Issue by cdrage** [redhat-developer/odo#899]
We should add the feature to be able to specify a specific branch from the Git directory when deploying.

For example:

```sh
odo create openjdk18 --git https://github.com/openshift-evangelists/Wild-West-Backend:foobar
# or
odo create openjdk18 --git https://github.com/openshift-evangelists/Wild-West-Backend --branch foobar
```

**Comment by metacosm** [comment]
-1 for me. This is best left to `git` itself. `odo` is already becoming quite large in scope and should stay as focused as possible around its core features.

**Comment by jorgemoralespou** [comment]
This is part of it's core features, and we should be able to create applications the same way `oc new-app` does, although I would not think this is a priority, I agree this should be supported.

@cdrage in case of --git based components, I think we will eventually need to support:
- branches
- context-directories/path
- git repository credentials (This will probably need to be generated)

The third option here to me would be low priority, but the first 2 should be something not too complex to set up. 

@kadel any opinion on this?

**Comment by jorgemoralespou** [comment]
@metacosm why this is part of `odo` core features is because once the developer has finished the iterative development cycle, he can configure that component to be based on git, and hence go through the natural s2i build cycle.

But, we need to understand how would a developer work. Maybe he would just delete the application/component once he's done.

**Comment by kadel** [comment]
I guess this depends on what approach we choose. 
Right now we are assuming that git components will be used to deploy the application after the development is done.
If we continue this direction than it makes sense to add support for branches, paths, credentials etc...

BTW: Adding support for choosing a branch is going to be easy, something like adding 4 lines.

I think that there should be something better than just building component from git once the development process is done. But currently, I don't have an idea how it should look like :-(

**Comment by geoand** [comment]
+1 from me

**Comment by girishramnani** [comment]
if we add support for branching, shouldn't we consider specific commits or tags?. usually the approach I have seen is something like this  `https://example.com/vpc.git?ref=v1.2.0` where ref supports tags, branches and commits.

**Comment by metacosm** [comment]
> if we add support for branching, shouldn't we consider specific commits or tags?. usually the approach I have seen is something like this `https://example.com/vpc.git?ref=v1.2.0` where ref supports tags, branches and commits.

+1 if we add this feature, this should work with an arbitrary ref, not just branches

**Comment by kadel** [comment]
> if we add support for branching, shouldn't we consider specific commits or tags?. usually the approach I have seen is something like this `https://example.com/vpc.git?ref=v1.2.0` where ref supports tags, branches and commits.

agree it should be universal and take any ref.

Is `?ref` something that is standard or used anywhere else?

**Comment by jorgemoralespou** [comment]
I would rather add flags, that can easily be interpreted/completed by the interactive mode that add complex syntax.
Eventually, this would translate to a buildConfig in Openshift (https://docs.openshift.com/container-platform/3.9/dev_guide/builds/build_inputs.html#source-code), and they use `ref` as the internal name, but use `#` in the git notation to separate the value, which I've always found weird, given the other so many flags that there are? 

Is adding flags to the cli a problem? OpenShift engineers have always been averse to adding new flags, but I've never know why :-D

**Comment by kadel** [comment]
I don't have anything against flags.
I expected that you will be against it :-D as you always prefer options that require less typing :-)

Let's use `--ref` flag. It will allow us to use one flag for tag/branch/commit hash.

We could also add separate flags `--branch`, `--tag`, `--commit` but that would be just 3 flags that do exactly the same thing.

**Comment by girishramnani** [comment]
+1 on that

**Comment by jorgemoralespou** [comment]
> I expected that you will be against it :-D as you always prefer options that require less typing :-)

This is a corner case. I like less going in common scenarios. Else, supported via flags. How many times a developer will use this flag? That's usually the question.

So, +1 to `--ref` and prior documentation that it supports branch and tag names.

**Comment by cdrage** [comment]
The `--ref`, `--branch`, `--tag` and `--commit` flags all LGTM. 

@girishramnani Want to proceed with the implementation?

**Comment by girishramnani** [comment]
Sure.

**Comment by kadel** [comment]
Ok, let's just add one flag -  `--ref`. Documentation should explain that this can be used for any git reference (branch, tag, commit).

**Comment by girishramnani** [comment]
blocked until https://github.com/redhat-developer/odo/pull/991 gets merged

**Comment by geoand** [comment]
#991 has been merged

**Comment by girishramnani** [comment]
@cdrage shouldn't we consider the update scenerio as well e.g. ` odo update --git https://github.com/openshift/nodejs-ex.git --ref v0.0.1` ?

**Comment by cdrage** [comment]
@girishramnani Yes, we'll need to update this for `odo update` as well.
